include/cuco/detail/storage.cuh

```diff
@@ -0,0 +1,413 @@
+
+  using value_type  = T;                                          ///< Type of structs
+  using window_type = cuda::std::array<value_type, window_size>;  ///< Type of struct windows
+
+  /**
+   * @brief Constructor of AoW base storage.
+   *
+   * @param size Number of elemets to store
+   */
+  explicit constexpr aow_storage_base(Extent size) : storage_base<Extent>{size} {}
+
+  /**
+   * @brief Gets the total number of slot windows in the current storage.
+   *
+   * @return The total number of slot windows
+   */
+  [[nodiscard]] __host__ __device__ inline constexpr extent_type num_windows() const noexcept
+  {
+    return storage_base<Extent>::capacity();
+  }
+
+  /**
+   * @brief Gets the total number of slots in the current storage.
+   *
+   * @return The total number of slots
+   */
+  [[nodiscard]] __host__ __device__ inline constexpr auto capacity() const noexcept
+  {
+    return storage_base<Extent>::capacity().template multiply<window_size>();
+  }
+};
+
+/**
+ * @brief Non-owning AoW storage reference type.
+ *
+ * @tparam WindowSize Number of slots in each window
+ * @tparam T Storage element type
+ * @tparam Extent Type of extent denoting storage capacity
+ */
+template <int32_t WindowSize, typename T, typename Extent>
+class aow_storage_ref : public aow_storage_base<WindowSize, T, Extent> {
+ public:
+  using base_type = aow_storage_base<WindowSize, T, Extent>;  ///< AoW base class type
+
+  using base_type::window_size;  ///< Number of elements processed per window
+
+  using extent_type = typename base_type::extent_type;  ///< Storage extent type
+  using size_type   = typename base_type::size_type;    ///< Storage size type
+  using value_type  = typename base_type::value_type;   ///< Storage element type
+  using window_type = typename base_type::window_type;  ///< Window storage type
+
+  using base_type::capacity;
+  using base_type::num_windows;
+
+  /**
+   * @brief Constructor of AoS storage ref.
+   *
+   * @param windows Pointer to the windows array
+   * @param num_windows Number of slots
+   */
+  explicit constexpr aow_storage_ref(Extent num_windows, window_type* windows) noexcept
+    : aow_storage_base<WindowSize, T, Extent>{num_windows}, windows_{windows}
+  {
+  }
+
+  /**
+   * @brief Gets windows array.
+   *
+   * @return Pointer to the first window
+   */
+  [[nodiscard]] __device__ inline constexpr window_type* windows() noexcept { return windows_; }
+
+  /**
+   * @brief Gets windows array.
+   *
+   * @return Pointer to the first window
+   */
+  [[nodiscard]] __device__ inline constexpr window_type* windows() const noexcept
+  {
+    return windows_;
+  }
+
+  /**
+   * @brief Returns an array of elements (window) for a given index.
+   *
+   * @param index Index of the first element of the window
+   * @return An array of elements
+   */
+  [[nodiscard]] __device__ inline constexpr window_type window(size_type index) const noexcept
+  {
+    return *(windows_ + index);
```

Comment:
```suggestion
    return *reinterpret_cast<window_type*>(__builtin_assume_aligned(windows_ + index, sizeof(value_type)*window_size));
```

Are we sure this always emits vectorized loads?
The compiler isn't even able to prove the alignment by itself in this simple example: https://godbolt.org/z/YffTxhxq3